### ISSUE TITLE
Check failed: 0.0 <= deterministic_duration (0 vs. -9.02616e-19) 

### ISSUE THREAD

**Issue by Shizcow** [google/or-tools#4342]
**What version of OR-Tools and what language are you using?**
Version: v9.10
Language: C++
Solver: `SAT_INTEGER_PROGRAMMING`
OS: Linux (Ubuntu 22.04 LTS)

I'm deterministically getting an internal assert to trip that I believe shouldn't be tripping.
How I ran into this behavior:
- I set up a model
- I added hints to the model of a known good solution (unsure if this is related)
- I began an optimization problem
- Allow the solver 32 threads

And then I got the following dump:
```
Scaling to pure integer problem.
Num integers: 119168/119168 (implied: 0 in_inequalities: 0 max_scaling: 0) [IP] 
Maximum constraint coefficient relative error: 0
Maximum constraint worst-case activity error: 0
Constraint scaling factor range: [1, 1]

Starting CP-SAT solver v9.10.4067
Parameters: max_time_in_seconds: 180 log_search_progress: true keep_all_feasible_solutions_in_presolve: true num_workers: 32

Initial optimization model 'xbars': (model_fingerprint: 0x2c282de46cb516ec)
#Variables: 119'168 (#bools: 128 in floating point objective)
  - 119'168 Booleans in [0,1]
#kLinear1: 27'565'696
#kLinear2: 161'791
#kLinear3: 768
#kLinearN: 8'913 (#terms: 246'016)

Starting presolve at 3.79s
The solution hint is complete and is feasible.
[Scaling] Floating point objective has 128 terms with magnitude in [0.2, 0.2] average = 0.2
[Scaling] Objective coefficient relative error: 0
[Scaling] Objective worst-case absolute error: 0
[Scaling] Objective scaling factor: 5
  1.09e+00s  0.00e+00d  [PresolveToFixPoint] #num_loops=1 
  1.86e-03s  0.00e+00d  [ExtractEncodingFromLinear] #potential_supersets=465 
  1.28e+00s  1.00e+00d *[Probe] #probed=66'546 #equiv=50'560 #new_binary_clauses=33'273 
  3.83e+00s  0.00e+00d  [MaxClique] 
  1.81e-01s  0.00e+00d  [PresolveToFixPoint] 
  5.25e-02s  0.00e+00d  [ProcessAtMostOneAndLinear] 
  4.10e-02s  0.00e+00d  [DetectDuplicateConstraints] 
  1.44e-02s  2.53e-04d  [DetectDominatedLinearConstraints] #relevant_constraints=256 
  2.98e-03s  0.00e+00d  [DetectDifferentVariables] 
  3.21e+00s  1.83e-03d  [ProcessSetPPC] #relevant_constraints=70'352 #num_inclusions=60'544 
  2.01e-02s  0.00e+00d  [FindAlmostIdenticalLinearConstraints] 
  6.05e-01s  4.09e-01d  [FindBigAtMostOneAndLinearOverlap] 
  3.33e-02s  1.54e-01d  [FindBigVerticalLinearOverlap] 
  8.00e-02s  9.99e-02d  [FindBigHorizontalLinearOverlap] #linears=256 
  9.00e-03s  3.48e-04d  [MergeClauses] 
  7.47e-02s  0.00e+00d  [PresolveToFixPoint] 
  7.29e-02s  0.00e+00d  [PresolveToFixPoint] 
  1.22e+00s  1.01e+00d *[Probe] #probed=108'864 #new_binary_clauses=93'216 
  3.77e+00s  0.00e+00d  [MaxClique] Merged 60671(121342 literals) into 60159(121086 literals) at_most_ones. 
  1.53e-01s  0.00e+00d  [PresolveToFixPoint] 
  5.17e-02s  0.00e+00d  [ProcessAtMostOneAndLinear] 
  4.19e-02s  0.00e+00d  [DetectDuplicateConstraints] 
  1.48e-02s  2.53e-04d  [DetectDominatedLinearConstraints] #relevant_constraints=256 
  3.45e-03s  0.00e+00d  [DetectDifferentVariables] 
  3.20e+00s  1.82e-03d  [ProcessSetPPC] #relevant_constraints=69'840 #num_inclusions=60'032 
  1.97e-02s  0.00e+00d  [FindAlmostIdenticalLinearConstraints] 
  6.02e-01s  4.10e-01d  [FindBigAtMostOneAndLinearOverlap] 
  3.26e-02s  1.54e-01d  [FindBigVerticalLinearOverlap] 
  7.97e-02s  9.99e-02d  [FindBigHorizontalLinearOverlap] #linears=256 
  9.37e-03s  3.48e-04d  [MergeClauses] 
  7.68e-02s  0.00e+00d  [PresolveToFixPoint] 
  7.53e-02s  0.00e+00d  [PresolveToFixPoint] 
  1.22e+00s  1.01e+00d *[Probe] #probed=108'862 #new_binary_clauses=92'447 
  3.77e+00s  0.00e+00d  [MaxClique] 
  1.54e-01s  0.00e+00d  [PresolveToFixPoint] 
  5.24e-02s  0.00e+00d  [ProcessAtMostOneAndLinear] 
  4.27e-02s  0.00e+00d  [DetectDuplicateConstraints] 
  1.54e-02s  2.53e-04d  [DetectDominatedLinearConstraints] #relevant_constraints=256 
  3.90e-03s  0.00e+00d  [DetectDifferentVariables] 
  3.21e+00s  1.82e-03d  [ProcessSetPPC] #relevant_constraints=69'840 #num_inclusions=60'032 
  2.10e-02s  0.00e+00d  [FindAlmostIdenticalLinearConstraints] 
  6.00e-01s  4.09e-01d  [FindBigAtMostOneAndLinearOverlap] 
  3.34e-02s  1.54e-01d  [FindBigVerticalLinearOverlap] 
  8.10e-02s  9.99e-02d  [FindBigHorizontalLinearOverlap] #linears=256 
  9.85e-03s  3.48e-04d  [MergeClauses] 
  7.80e-02s  0.00e+00d  [PresolveToFixPoint] 
  4.76e-02s  0.00e+00d  [ExpandObjective] #entries=7'559'040 #tight_variables=59'520 #tight_constraints=465 

Presolve summary:
  - 50560 affine relations were detected.
  - rule 'TODO linear2: contains a Boolean.' was applied 323'582 times.
  - rule 'affine: new relation' was applied 50'560 times.
  - rule 'at_most_one: transformed into max clique.' was applied 1 time.
  - rule 'bool_or: implications' was applied 161'791 times.
  - rule 'linear: always true' was applied 27'565'696 times.
  - rule 'linear: fixed or dup variables' was applied 128 times.
  - rule 'linear: negative clause' was applied 111'104 times.
  - rule 'linear: positive clause' was applied 59'647 times.
  - rule 'linear: positive equal one' was applied 465 times.
  - rule 'linear: remapped using affine relations' was applied 128 times.
  - rule 'linear: simplified rhs' was applied 171'007 times.
  - rule 'presolve: 0 unused variables removed.' was applied 1 time.
  - rule 'presolve: iteration' was applied 3 times.

Presolved optimization model 'xbars': (model_fingerprint: 0xea84297d858dc462)
#Variables: 68'608 (#bools: 128 in objective)
  - 68'608 Booleans in [0,1]
#kAtMostOne: 768 (#literals: 2'304)
#kBoolAnd: 9'087 (#enforced: 9'087) (#literals: 68'478)
#kBoolOr: 8'960 (#literals: 69'504)
#kExactlyOne: 465 (#literals: 59'520)
#kLinearN: 256 (#terms: 78'080)

Preloading model.
#Bound  38.98s best:inf   next:[0,25.6]   initial_domain
#1      38.98s best:25.6  next:[0,25.4]   complete_hint
[Symmetry] Graph for symmetry has 216'657 nodes and 465'790 arcs.
[Symmetry] Symmetry computation done. time: 11.8788 dtime: 0.208826
[Symmetry] #generators: 342, average support size: 270.971
[Symmetry] Found orbitope of size 128 x 6
#Model  50.94s var:68608/68608 constraints:19536/19536

Starting search at 50.96s with 32 workers.
16 full problem subsolvers: [core, default_lp, lb_tree_search, max_lp, no_lp, objective_lb_search, objective_lb_search_max_lp, objective_lb_search_no_lp, objective_shaving_search_max_lp, objective_shaving_search_no_lp, probing, probing_max_lp, pseudo_costs, quick_restart, quick_restart_no_lp, reduced_costs]
14 first solution subsolvers: [fj_long_default, fj_long_lin_default, fj_long_random, fj_short_default, fj_short_lin_default, fj_short_lin_random, fj_short_random, fs_random(2), fs_random_no_lp(2), fs_random_quick_restart(2), fs_random_quick_restart_no_lp]
11 incomplete subsolvers: [feasibility_pump, graph_arc_lns, graph_cst_lns, graph_dec_lns, graph_var_lns, rins/rens, rnd_cst_lns, rnd_var_lns, violation_ls(3)]
3 helper subsolvers: [neighborhood_helper, synchronization_agent, update_gap_integral]
#Bound  51.16s best:25.6  next:[0.2,25.4] objective_shaving_search_max_lp (vars=68608 csts=19537)
#2      51.28s best:25.4  next:[0.2,25.2] violation_ls(batch:1 #solutions_imported:1 #lin_moves:18 #lin_evals:734 #weight_updates:9)
#3      51.68s best:25.2  next:[0.2,25]   violation_ls(batch:3 #solutions_imported:2 #lin_moves:2'840 #lin_evals:326'179 #gen_moves:1'450 #gen_evals:0 #comp_moves:0 #backtracks:724 #weight_updates:2'862)
#4      52.96s best:24.8  next:[0.2,24.6] violation_ls(batch:4 #solutions_imported:3 #lin_moves:3'201 #lin_evals:337'932 #gen_moves:1'450 #gen_evals:0 #comp_moves:0 #backtracks:724 #weight_updates:3'227)
#5      52.97s best:23.8  next:[0.2,23.6] violation_ls(batch:4 #solutions_imported:3 #lin_moves:18 #lin_evals:201'011 #gen_moves:917 #gen_evals:0 #comp_moves:7 #backtracks:455 #weight_updates:9)
#6      52.98s best:23.6  next:[0.2,23.4] violation_ls(batch:4 #solutions_imported:2 #lin_moves:1 #lin_evals:261'206 #gen_moves:2'258 #gen_evals:0 #comp_moves:31 #backtracks:1'112 #weight_updates:8)
#Bound  53.10s best:23.6  next:[0.4,23.4] quick_restart_no_lp
#Model  53.13s var:62712/68608 constraints:17897/19536
#Model  53.19s var:62710/68608 constraints:17895/19536
#7      53.37s best:23.4  next:[0.4,23.2] violation_ls(batch:5 #solutions_imported:4 #lin_moves:3'201 #lin_evals:337'932 #gen_moves:1'450 #gen_evals:0 #comp_moves:0 #backtracks:724 #weight_updates:3'227)
F0000 00:00:1723237904.734763 1986256 time_limit.h:179] Check failed: 0.0 <= deterministic_duration (0 vs. -9.02616e-19) 
*** Check failure stack trace: ***
Aborted (core dumped)
```

I'm at a complete loss on how to debug this one. Any thoughts? I've been developing on this project for a few months and this popped up when I made the following change to my project structure:
- Previously, I would solve for an initial solution, hint the solver, and then do optimization (in a separate `solve` call). This worked fine.
- I implemented a caching method to save the initial solution to disk and recall.
- I set up cache read as a hint into the initial solution solver call. This also worked fine, but was slow because I had to wait for the solver to validate the hint.
- I then validated the cached read in C++ and skipped over the initial solver call. AKA, only hinting and then jumping straight into optimization. Then I get the crash.

Running under GDB fails to repro, so I can't grab a stacktrace. Failure occurs with 31, 32 and 33 threads every time, deterministically. With 60 or 16 threads I don't run into this issue (I can test more on request). So I do have a workaround, at least.

FWIW, I've compiled with GCC on march=native, LTO, and O3. I can test with other optimizations if someone believes this may be a GCC bug.

I'm pretty new to using OR-Tools, so if there's a way I can dump a repro from this, let me know and I'd be happy to upload.

**Comment by lperron** [comment]
fixed on main.

**Comment by Shizcow** [comment]
I've confirmed that my specific issue has been fixed on main as of 39b07a9188800edf62dd1fc279eee463443d40d1. Thanks @lperron!

**Comment by lperron** [comment]
thanks for checking.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le lun. 12 août 2024 à 08:06, Devin Pohl ***@***.***> a
écrit :
